### DISCUSSION TITLE
rule 的两种使用方式的差异

### DISCUSSION THREAD

**Discussion by q962** [xmake-io/xmake#5434] [Q&A]
- `add_files("", {rule = "my_rule"}`
只会导致以下函数被调用
```
before_build_files
on_build_files
after_build_files
```
这种丢失 on_load / on_config 应该是个错误吧

- `add_rules("my_rule");`
只会导致以下函数被调用
```
on_load
on_config
before_build
on_build
after_build
```

**Comment by waruqi** [comment]
add_files 里应用 rules 是有限制的，如果你想执行 on_load/on_config 建议走 add_rules + set_extensions
